### PR TITLE
Auth middleware

### PR DESCRIPTION
**Pull request by Enochen**
### Summary

This extends the Express `req` object/type with a field `authInfo` containing firebase authentication info.

It's up to each endpoint handler function to check `req.authInfo` if necessary. The field will be falsy if there was no valid authentication provided.

`authInfo` is of type [DecodedIdToken](https://firebase.google.com/docs/reference/admin/node/admin.auth.DecodedIdToken), which has fields like `email`.
This PR doesn't include a check for cornell.edu emails.

#### Client Usage
Authentication is provided by the client through the headers, in the format:
`Authentication: Bearer [id]`
image

The `id` would be a Firebase Auth ID token JWT string, which clients get through the Firebase Sign-in API.

### Testing

Tested with Postman with both valid and invalid tokens

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -41,7 +41,7 @@
     // "baseUrl": "./",                       /* Base directory to resolve non-absolute module names. */
     // "paths": {},                           /* A series of entries which re-map imports to lookup locations relative to the 'baseUrl'. */
     // "rootDirs": [],                        /* List of root folders whose combined content represents the structure of the project at runtime. */
-    // "typeRoots": [],                       /* List of folders to include type definitions from. */
+    "typeRoots": ["/"],                       /* List of folders to include type definitions from. */
```
Comment (SamChou19815) [inline tsconfig.json:44]: Why do we need this?

File: app/app.ts [hunk 2]
```diff
@@ -9,64 +9,83 @@ import * as userHandler from "./handlers/userHandler";
...
-      res.status(400).json(error);
-    });
+    promise
+      .then((val) => {
+        res.status(200).send(val);
+      })
+      .catch((reason) => {
+        let error = { error: "Promise rejected for: " + reason };
+        res.status(400).json(error);
+      });
```
Comment (SamChou19815) [inline app/app.ts]: Avoid these format-only changes

File: app/app.ts [hunk 3]
```diff
@@ -9,64 +9,83 @@ import * as userHandler from "./handlers/userHandler";
...
+    shell(orgHandler, orgHandler.getOrg, false, req, res, [db, req, res])
+  );
+  app.post("/updateOrg/:orgID", (req, res) =>
+    shell(orgHandler, orgHandler.updateOrg, false, req, res, [db, req, res])
+  );
 
-  app.listen(port, () => console.log(`Backend running on http://localhost:${port}`))
+  app.listen(port, () =>
+    console.log(`Backend running on http://localhost:${port}`)
+  );
```
Comment (SamChou19815) [inline app/app.ts]: Avoid format-only changes

File: app/handlers/orgHandler.ts [hunk 4]
```diff
@@ -24,7 +24,7 @@ export const createOrg = async (
       return { orgId : orgRef.id };
     }).catch(
     async error => {
-     return { error : error} ;
+      return { error : error };
```
Comment (SamChou19815) [inline app/handlers/orgHandler.ts]: Avoid format only changes

**Review by SamChou19815 — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 1 above)
Comment (Enochen) [reply tsconfig.json:44]: It tells tsnode to look for the types.d.ts file which extends the Request interface.

a TSError is thrown otherwise

Comment (SamChou19815) [reply tsconfig.json:44]: This is weird, TS should pick that file up automatically.

File: app/handlers/orgHandler.ts (hunk 4 above)
Comment (Enochen) [reply app/handlers/orgHandler.ts]: Ok, should we put formatting changes in a separate PR next time?

File: tsconfig.json (hunk 1 above)
Comment (Enochen) [reply tsconfig.json:44]: Yeah I was confused too but this option seems to be the solution

**Review by SamChou19815 — APPROVED** [review]

**Pull request merged**